### ISSUE TITLE
Package request: hid-fanatecff

### ISSUE THREAD

**Issue by rake5k** [NixOS/nixpkgs#271117]
**Project description**

Linux module driver for Fanatec driving wheels.

**Metadata**

* homepage URL: https://github.com/gotzl/hid-fanatecff
* source URL: https://github.com/gotzl/hid-fanatecff
* license: gpl2
* platforms: linux

**Priorities**

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by PhilT** [comment]
@rake5k Have you got a working config for this? If not I'm working on it.

**Comment by rake5k** [comment]
Not at all, thank you a lot for looking into it! But I am glad to help out with testing if it would help.

**Comment by PhilT** [comment]
I've never packaged anything in NixOS before (Only been using it for a couple of months) but I managed to get a derivation working locally.

```
❯ cat src/hid-fanatecff/default.nix

{ lib
, stdenv
, fetchFromGitHub
, kernel
, kmod
, linuxConsoleTools
}:

let moduledir = "lib/modules/${kernel.version}/kernel/drivers/hid";
in
stdenv.mkDerivation rec {
  pname = "hid-fanatecff";
  version = "0.0.2";
  name = "hid-fanatecff-${version}-${kernel.version}";

  src = fetchFromGitHub {
    owner = "gotzl";
    repo = "hid-fanatecff";
    rev = "next";
    sha256 = "hHxh0VFHwLyAxUkXSpjc2VlXdyoABoenc2sAsewhLhs=";
  };

  hardeningDisable = [ "pic" "format" ];
  nativeBuildInputs = kernel.moduleBuildDependencies;

  patchPhase = ''
    mkdir -p $out/lib/udev/rules.d
    mkdir -p $out/${moduledir}
    substituteInPlace Makefile --replace "/etc/udev/rules.d" "$out/lib/udev/rules.d"
    substituteInPlace fanatec.rules --replace "/usr/bin/evdev-joystick" "${linuxConsoleTools}/bin/evdev-joystick" --replace "GROUP:=\"plugdev\"" "TAG+=\"uaccess\""
    sed -i '/depmod/d' Makefile
  '';

  makeFlags = [
    "KVERSION=${kernel.modDirVersion}"
    "KDIR=${kernel.dev}/lib/modules/${kernel.modDirVersion}/build"
    "MODULEDIR=$(out)/${moduledir}"
  ];
}
```

```
❯ cat src/configuration.nix
{ config, pkgs, ... }:

let fanatecff = pkgs.linuxPackages.callPackage ./hid-fanatecff/default.nix {};
in
{
  boot.extraModulePackages = [ fanatecff ];
  services.udev.packages = [ fanatecff ];
  boot.kernelModules = [ "hid-fanatec" ];

  environment.systemPackages = with pkgs; [
  ];
}
```

Full repo is [here](https://github.com/PhilT/nixos-files) and that `configuration.nix file is just part of a bigger config [here](https://github.com/PhilT/nixos-files/blob/main/src/spruce/default.nix).

I will be taking a look into packaging it into nixpkgs when I get the chance.

**Comment by PhilT** [comment]
I'm going to hold off adding this to nixpkgs for now until more people here request it. Happy to do it once more :+1: are added.

**Comment by PhilT** [comment]
I've updated my nix config to work with the latest hid-fanatecff
https://github.com/PhilT/nixos-files/blob/main/src/hid-fanatecff/default.nix

**Comment by sigonze** [comment]
> I've updated my nix config to work with the latest hid-fanatecff https://github.com/PhilT/nixos-files/blob/main/src/hid-fanatecff/default.nix

I have tested it and it is working for me (I only updated to the latest revision).
I only had to use the default kernel instead of xanmod to avoid an error during kernel version check. Since I am a newcomer on NixOS I did not figure out yet how to deal wih it yet.
Thanks for the sharing.

**Comment by sigonze** [comment]
Finally I was able to make it work with xanmod. The solution is to replace `${kernel.version}` by `${kernel.modDirVersion}`.
https://github.com/sigonze/nixos-config/blob/main/hid-fanatecff/default.nix
For udev rules, `games` group has to be created and `linuxConsoleTools` has to be installed.

**Comment by Zombiefleischer** [comment]
What is the current status with this package?

**Comment by PhilT** [comment]
I don't have Fanatec kit anymore so I'm not working on this. At the time there wasn't much interest in getting it into Nixpkgs. However, I was using it successfully as a local derivation. Grab either my code above or use the updated one from @sigonze and you should be good to go.

**Comment by sigonze** [comment]
Still using the code from @PhilT my side and it is still ok. (for instance it worked perfectly on Assetto Corsa Evo yesterday)
I only made few changes to the original code. Only to activate it using an option and to extract the version in json file to make it more convenient for updates.
It is available here: https://github.com/sigonze/nixos-config/tree/main/src/drivers
What is needed are the fanatec.nix file and hid-fanatecff folder.
Hope it will help.
